### PR TITLE
Fixes/file open picker skia wpf

### LINKED ISSUE #7683: [WPF] FileOpenPicker System.ArgumentException
### Current behavior

Exception occurred: System.ArgumentException: Provided filter string is not valid. Filter string should contain a description of the filter, followed by a vertical bar and the filter pattern. Must also separate multiple filter description and pattern pairs by a vertical bar. Must separate multiple extensions in a filter pattern with a semicolon. Example: \"Image files (*.bmp, *.jpg)|*.bmp;*.jpg|All files (*.*)|*.*\"

### Expected behavior

_No response_

### How to reproduce it (as minimally and precisely as possible)

1. Run SamplesApp.Skia.WPF
2. Go to FileOpenPickerTests (Windows.Storage -> FileOpenPickerTests)
3. Add File Type
4. Click on Pick Single o Multiple File/s

### Workaround

_No response_

### Works on UWP/WinUI

Yes

### Environment

Uno.UI / Uno.UI.WebAssembly / Uno.UI.Skia

### NuGet package version(s)

4.1.0-dev.160

### Affected platforms

Skia (GTK on Linux/macOS/Windows)

### IDE

Visual Studio 2022

### IDE version

17.1.0 Preview 1.1

### Relevant plugins

_No response_

### Anything else we need to know?

_No response_

### LINKED ISSUE #7685: [WPF] FileOpenPicker duplicate File Type *
### Current behavior

if add code like this
```csharp
        fileOpenPicker.FileTypeFilter.Add(".xls");
        fileOpenPicker.FileTypeFilter.Add(".xlsx");
        fileOpenPicker.FileTypeFilter.Add("*");
```

### WPF
immagine

### UWP
immagine

### Expected behavior

Show as UWP
immagine

### How to reproduce it (as minimally and precisely as possible)

_No response_

### Workaround

_No response_

### Works on UWP/WinUI

Yes

### Environment

Uno.UI / Uno.UI.WebAssembly / Uno.UI.Skia

### NuGet package version(s)

4.1.0-dev.160

### Affected platforms

Skia (GTK on Linux/macOS/Windows)

### IDE

Visual Studio 2022

### IDE version

17.1.0 Preview 1.1

### Relevant plugins

_No response_

### Anything else we need to know?

_No response_

### PR DESCRIPTION
**Pull request by workgroupengineering**
GitHub Issue (If applicable): closes #

fixes #7683
fixes #7685 

## PR Type

What kind of change does this PR introduce?

- Bugfix

## What is the new behavior?

Restore same behavior as UWP FileOpenPicker on Skia.WPF

## PR Checklist

Please check if your PR fulfills the following requirements:

- [ ] Docs have been added/updated which fit [documentation template](https://github.com/unoplatform/uno/blob/master/doc/.feature-template.md) (for bug fixes / features)
- [ ] [Unit Tests and/or UI Tests](https://github.com/unoplatform/uno/blob/master/doc/articles/uno-development/working-with-the-samples-apps.md) for the changes have been added (for bug fixes / features) (if applicable)
- [ ] Validated PR `Screenshots Compare Test Run` results.
- [x] Contains **NO** breaking changes
- [x] Associated with an issue (GitHub or internal) and uses the [automatic close keywords](https://help.github.com/en/github/managing-your-work-on-github/linking-a-pull-request-to-an-issue).
- [x] Commits must be following the [Conventional Commits](https://www.conventionalcommits.org/en/v1.0.0/#summary) specification.

## Other information

Internal Issue (If applicable):

### CODE REVIEW

File: src/Uno.UI.Runtime.Skia.Wpf/Extensions/Storage/Pickers/FileOpenPickerExtension.cs [hunk 1]
```diff
@@ -45,17 +45,34 @@ private async Task<StorageFile[]> OpenPickerAsync(bool multiple)
 			};
 
 			var filterBuilder = new StringBuilder();
-			filterBuilder.Append(string.Join("|", _picker.FileTypeFilter.Select(fileType => $"{fileType}|*{fileType}")));
+			var all = false;
+			var filterIndex = 0;
+			foreach (var fileType in _picker.FileTypeFilter)
+			{
+				if (fileType == "*")
```
Comment (Youssef1313) [inline src/Uno.UI.Runtime.Skia.Wpf/Extensions/Storage/Pickers/FileOpenPickerExtension.cs:52]: I think this doesn't match UWP behavior when multiple filter types are all `*`

But it's probably not important.

File: src/Uno.UI.Runtime.Skia.Wpf/Extensions/Storage/Pickers/FileOpenPickerExtension.cs [hunk 2]
```diff
@@ -45,17 +45,34 @@ private async Task<StorageFile[]> OpenPickerAsync(bool multiple)
...
+				{
+					filterBuilder.Append($"|{fileType}|*{fileType}");
+					filterIndex++;
+				}
+			}
 
-			if (_picker.FileTypeFilter.Count > 1)
+			if (all)
+			{
+				filterBuilder.Append("|All|*");
```
Comment (Youssef1313) [inline src/Uno.UI.Runtime.Skia.Wpf/Extensions/Storage/Pickers/FileOpenPickerExtension.cs:65]: image

This is not necessarily the last item. But we need to keep track of its index to set `FilterIndex` properly.

**Review by Youssef1313 — COMMENTED** [review]
I think the following more matches UWP behavior. But it's not tested and I could be wrong.

```csharp
			var allIndex = -1;
			var currentIndex = -1;
			foreach (var fileType in _picker.FileTypeFilter)
			{
				currentIndex++;
				if (fileType == "*")
				{
					filterBuilder.Append("|All|*");
					allIndex = currentIndex;
				}
				else
				{
					filterBuilder.Append($"|{fileType}|*{fileType}");
				}
			}

			if (allIndex == -1)
			{
				var fullFilter = string.Join(";", _picker.FileTypeFilter.Select(fileType => $"*{fileType}"));
				filterBuilder.Append($"|All|{fullFilter}");
				allIndex = currentIndex + 1;
			}

			openFileDialog.Filter = filterBuilder.ToString(1, filterBuilder.Length - 1);
			openFileDialog.FilterIndex = allIndex;
```

File: src/Uno.UI.Runtime.Skia.Wpf/Extensions/Storage/Pickers/FileOpenPickerExtension.cs (hunk 1 above)
Comment (workgroupengineering) [reply src/Uno.UI.Runtime.Skia.Wpf/Extensions/Storage/Pickers/FileOpenPickerExtension.cs:52]: This is  UWP behavior
immagine

Comment (Youssef1313) [reply src/Uno.UI.Runtime.Skia.Wpf/Extensions/Storage/Pickers/FileOpenPickerExtension.cs:52]: @workgroupengineering I mean when having multiple `*` specified. This is how UWP behaves:

image

File: src/Uno.UI.Runtime.Skia.Wpf/Extensions/Storage/Pickers/FileOpenPickerExtension.cs (hunk 2 above)
Comment (workgroupengineering) [reply src/Uno.UI.Runtime.Skia.Wpf/Extensions/Storage/Pickers/FileOpenPickerExtension.cs]: Thanks I hadn't considered this scenario

File: src/Uno.UI.Runtime.Skia.Wpf/Extensions/Storage/Pickers/FileOpenPickerExtension.cs [hunk 3]
```diff
@@ -45,17 +45,31 @@ private async Task<StorageFile[]> OpenPickerAsync(bool multiple)
...
-			filterBuilder.Append(string.Join("|", _picker.FileTypeFilter.Select(fileType => $"{fileType}|*{fileType}")));
+			var filterIndex = -1;
+			var fileTypeFilterCount = _picker.FileTypeFilter.Count;
+			for (var i = 0; i < fileTypeFilterCount; i++)
+			{
+				var fileType = _picker.FileTypeFilter[i];
+				if (fileType == "*")
+				{
+					filterBuilder.Append("|All|*");
+					filterIndex = i + 1;
```
Comment (Youssef1313) [inline src/Uno.UI.Runtime.Skia.Wpf/Extensions/Storage/Pickers/FileOpenPickerExtension.cs:56]: The `+ 1` here is confusing me. Don't we want the same index as the `*` ?

Comment (workgroupengineering) [reply src/Uno.UI.Runtime.Skia.Wpf/Extensions/Storage/Pickers/FileOpenPickerExtension.cs:56]: Microsoft.Win32.OpenFileDialog.FilterIndex  is 1-based see [this](https://docs.microsoft.com/it-it/dotnet/api/microsoft.win32.filedialog.filterindex?view=windowsdesktop-6.0#remarks)

**Review by Youssef1313 — APPROVED** [review]
LGTM

File: src/Uno.UI.Runtime.Skia.Wpf/Extensions/Storage/Pickers/FileOpenPickerExtension.cs (hunk 1 above)
Comment (carldebilly) [reply src/Uno.UI.Runtime.Skia.Wpf/Extensions/Storage/Pickers/FileOpenPickerExtension.cs:53]: I think it's more `*.*` on Windows for _all files_, no?  Here's a screenshot from my NotePad:
image

Comment (Youssef1313) [reply src/Uno.UI.Runtime.Skia.Wpf/Extensions/Storage/Pickers/FileOpenPickerExtension.cs:53]: @carldebilly This is not how it behaves on UWP I think.

Comment (workgroupengineering) [reply src/Uno.UI.Runtime.Skia.Wpf/Extensions/Storage/Pickers/FileOpenPickerExtension.cs:53]: > @workgroupengineering I mean when having multiple `*` specified. This is how UWP behaves:
> 
> image
> 
> It's meaningless to specify `*` multiple times, that's why it's not much important.

It sounds strange, but in UWP the behavior is the same, I have checked now.



> @carldebilly This is not how it behaves on UWP I think.
I confirm

Comment (Youssef1313) [reply src/Uno.UI.Runtime.Skia.Wpf/Extensions/Storage/Pickers/FileOpenPickerExtension.cs:53]: > It sounds strange, but in UWP the behavior is the same, I have checked now.

Yup with the latest commits this got fixed. The initial commit when I reviewed didn't match UWP for this weird edge case.

**Review by MartinZikmund — APPROVED** [review]
👍 Nice PR!

File: src/Uno.UI.Runtime.Skia.Wpf/Extensions/Storage/Pickers/FileOpenPickerExtension.cs (hunk 3 above)
Comment (robloo) [reply src/Uno.UI.Runtime.Skia.Wpf/Extensions/Storage/Pickers/FileOpenPickerExtension.cs:56]: That would be a great in-code comment.

**Comment by jeromelaban** [comment]
Thanks for your contribution!

**Pull request merged**